src/inmanta/agent/executor.py

```diff
@@ -284,13 +326,32 @@ class VirtualEnvironmentManager:
     def __init__(self, envs_dir: str) -> None:
         self._environment_map: dict[EnvBlueprint, ExecutorVirtualEnvironment] = {}
         self.envs_dir: str = envs_dir
+        # We rely on a Named lock (`self._locks`) to be able to lock specific entries of the `_environment_map` dict. This
+        # allows us to prevent creating and deleting the same venv at a given time. The keys of this named lock are the hash of
+        # venv
         self._locks: NamedLock = NamedLock()
+        self._cleanup_scheduler = util.Scheduler("venv_cleanup_scheduler")
+
+    async def start(self) -> None:
+        interval = datetime.timedelta(days=1).total_seconds()
+        executor_venv_retention_time = cfg.executor_venv_retention_time.get()
+        assert datetime.timedelta(days=executor_venv_retention_time).total_seconds() > interval, (
+            "The `executor-venv-retention-time` should be larger than the period " "with which the executor touches its file!"
+        )
+
+        self._cleanup_scheduler.add_action(
+            action=self.clean_virtual_environments,
+            schedule=util.IntervalSchedule(
+                interval=interval,
+            ),
+        )
 
     def get_or_create_env_directory(self, blueprint: EnvBlueprint) -> tuple[str, bool]:
         """
         Retrieves the directory path for a virtual environment based on the given blueprint.
-        If the directory does not exist, it creates a new one. This method ensures that each
-        virtual environment has a unique storage location.
+        If the directory does not exist, it creates a new one. This method ensures that each virtual environment has a unique
+        storage location. This method must be executed under the self._locks.get(<blueprint-hash>) lock to ensure thread-safe
+        operations for each unique blueprint
 
         :param blueprint: The blueprint of the environment for which the storage is being determined.
         :return: A tuple containing the path to the directory and a boolean indicating whether the directory was newly created.
```

Comment:
I would add a comment to line 360 stating that the folder name has to be the hash, otherwise locking breaks in `clean_virtual_environments`